### PR TITLE
Avoid naming conflicts by installing shared objects to lib/nest instead of lib

### LINKED ISSUE #1195: Please don't fail when CMAKE_INSTALL_PREFIX is /usr/local
https://github.com/nest/nest-simulator/blob/master/CMakeLists.txt#L104

There's nothing wrong with installing into ```/usr/local```.
It normally installs into a stage directory when package is built.

### LINKED ISSUE #1222: Add `LD_LIBRARY_PATH` to guide for "writing an extension module"
To build an extension module for NEST, I followed the guide on the NEST Developer Space: https://nest.github.io/nest-simulator/extension_modules

However, after following steps 1-5 for installing the example module, I couldn't load it, but received this error: 
```Jun 06 09:23:59 Install [Error]: DynamicModuleManagementError
    Module 'mymodule' could not be opened.
    The dynamic loader returned the following error: 'file not found'.
```

I could easily fix this by adding the NEST lib folder to my `LD_LIBRARY_PATH`, like so:
`export LD_LIBRARY_PATH=~/opt/nest/ins/lib/nest`

Now, perhaps it would be a good idea to add this simple step to the documentation?

### PR DESCRIPTION
**Pull request by clinssen**
Fixes #1222 
Fixes #1195

Currently, shared objects (.so files) generated during the NEST Simulator build process are installed to `${CMAKE_INSTALL_PREFIX}/lib`. This behaviour is dangerous because it can potentially overwrite existing system libraries. For example, in case the prefix is `/usr` or `~/.local`, any file by the name `librandom.so`, `libprecise.so`, `libmodels.so` and so on, that is already present in `/usr/lib` or `~/.local/lib`, will be overwritten. I heard that a naming conflict did indeed happen in the past with an Infiniband library named `libtopology.so`, which overlaps with the NEST shared object by the same name.

Generally, libraries are expected to use a unique prefix for their files (in our case, this would be "nest", so our files would be `libnesttopology.so`, `libnestrandom.so`, etc.) Other libraries (see `ls /usr/lib`) seem to prefer to create their own subdirectories. This means that we would have one subdirectory called `nest`, under which we would place all .so files generated during the build (`libtopology.so`, `librandom.so`, ...) In fact, this directory already exists, but is currently only used for user modules: the MyModule example, as well as NESTML generated modules are installed there. 

This PR proposes to install *all but one* .so file in `lib/nest` to avoid potential naming collisions with existing files in the `lib` directory—the PyNEST .so file will be unaffected and stay in `lib/python3.x/site-packages/nest`.

Thanks to @Silmathoron for initially pointing this out! (See nest/nestml#480).

Notes for reviewers: 

* I chose to keep `${CMAKE_INSTALL_LIBDIR}` equal to `lib` rather than set it to `lib/nest`, because of the PyNEST library that is installed in `lib/python3.x/site-packages/nest`, i.e. *not* under `lib/nest`.
* Should we remove `NEST_MODULE_PATH` entirely, as it is now equal to the path for all other shared objects, and thus already in the rpath/LD_LIBRARY_PATH?

### CODE REVIEW

**Comment by sarakonradi** [comment]
@clinssen Thanks a lot for your work. Following @jougs's comment in #1222, I have been working on the documentation of ``nest_vars.sh`` and somehow missed this your PR or yours. Sorry if #1515 caused any confusion! 

I am happy pausing the work in #1515 until a decision is made here. We can then update the documentation accordingly in order to reflect any potential changes to ``NEST_MODULE_PATH``.

**Comment by clinssen** [comment]
No problem! I don't think there is any overlap in content of the PRs. We should indeed first decide whether or not to keep NEST_MODULE_PATH and then go from there.

File: cmake/ConfigureSummary.cmake [hunk 1]
```diff
@@ -188,6 +188,9 @@ function( NEST_PRINT_CONFIG_SUMMARY )
   message( "The NEST executable will be installed to:" )
   message( "  ${CMAKE_INSTALL_FULL_BINDIR}/" )
   message( "" )
+  message( "Shared objects (.so files) will be installed to:" )
```
Comment (jougs) [inline cmake/ConfigureSummary.cmake:191]:
```suggestion
  message( "NEST dynamic libraries and user modules will be installed to:" )
```
The rationale is that "shared objects" is a pretty low-level technical term and the files end in `.dylib` on macOS and `.dll` on Windows.

File: extras/nest-config.in [hunk 2]
```diff
@@ -66,7 +66,7 @@ while test $# -gt 0; do
         echo "@ALL_CXXFLAGS@"
         ;;
     --libs)
-        echo "-L$prefix/@CMAKE_INSTALL_LIBDIR@ @ALL_LIBS@"
+        echo "-L$prefix/@CMAKE_INSTALL_LIBDIR@/nest @ALL_LIBS@"
```
Comment (jougs) [inline extras/nest-config.in:69]: Shouldn't there be a similar change for the `--libdir` switch in [line 91](https://github.com/nest/nest-simulator/pull/1513/files#diff-07b895965dd0b3a757363159d0466805R91) of this file?

File: extras/nest_vars.sh.in [hunk 3]
```diff
@@ -9,14 +9,18 @@ export NEST_DATA_DIR="${NEST_INSTALL_DIR}/@CMAKE_INSTALL_DATADIR@"
 # NEST finds help files $NEST_DOC_DIR/help
 export NEST_DOC_DIR="${NEST_INSTALL_DIR}/@CMAKE_INSTALL_DOCDIR@"
 
-# The path where NEST looks for user modules.
-export NEST_MODULE_PATH="${NEST_INSTALL_DIR}/@CMAKE_INSTALL_LIBDIR@/nest"
+# The path where NEST shared libraries are generated. (Works also if $(DY)LD_LIBRARY_PATH is undefined.)
```
Comment (jougs) [inline extras/nest_vars.sh.in:12]:
```suggestion
# The path in which NEST dynamic libraries are found. Works also if $(DY)LD_LIBRARY_PATH is undefined.
```

File: extras/nest_vars.sh.in [hunk 4]
```diff
@@ -9,14 +9,18 @@ export NEST_DATA_DIR="${NEST_INSTALL_DIR}/@CMAKE_INSTALL_DATADIR@"
...
+else
+    export LD_LIBRARY_PATH="$NEST_INSTALL_DIR/@CMAKE_INSTALL_LIBDIR@/nest"${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}
+fi
 
 # The path where the PyNEST bindings are installed.
 export NEST_PYTHON_PREFIX="${NEST_INSTALL_DIR}/@PYEXECDIR@"
 
-# Prepend NEST to PYTHONPATH in a safe way even if PYTHONPATH is undefined
-export PYTHONPATH="${NEST_PYTHON_PREFIX}${PYTHONPATH:+:$PYTHONPATH}"
+# Prepend NEST to PYTHONPATH. (Works also if $PYTHONPATH is undefined.)
```
Comment (jougs) [inline extras/nest_vars.sh.in:22]:
```suggestion
# Prepend NEST to PYTHONPATH. Works also if $PYTHONPATH is undefined.
```

File: extras/nest_vars.sh.in [hunk 5]
```diff
@@ -9,14 +9,18 @@ export NEST_DATA_DIR="${NEST_INSTALL_DIR}/@CMAKE_INSTALL_DATADIR@"
...
+export PYTHONPATH="$NEST_PYTHON_PREFIX${PYTHONPATH:+:$PYTHONPATH}"
 
-# Make nest / sli /... executables visible.
+# Make ``nest``/``sli``/etc. executables visible.
```
Comment (jougs) [inline extras/nest_vars.sh.in:25]:
```suggestion
# Make nest/sli/... executables visible.
```
I don't think that the markdown syntax helps here.

File: extras/travis_build.sh [hunk 6]
```diff
@@ -193,7 +193,7 @@ if [ "$xSTATIC_ANALYSIS" = "1" ]; then
     # Set the command line arguments for the static code analysis script and execute it.
 
     # The names of the static code analysis tools executables.
-    VERA=vera++                   
+    VERA=vera++
```
Comment (jougs) [inline extras/travis_build.sh:196]: :heart:

File: cmake/ConfigureSummary.cmake (hunk 1 above)
Comment (clinssen) [reply cmake/ConfigureSummary.cmake:191]: Good point!

**Comment by clinssen** [comment]
Thanks for the feedback! The reason that the value of ``${CMAKE_INSTALL_LIBDIR}`` is not modified, but instead in several places the subdirectory ``nest`` is appended, is because of the installation location of ``pynestkernel.so``:

``${CMAKE_INSTALL_LIBDIR}/python3.5/site-packages/nest/pynestkernel.so``
``${CMAKE_INSTALL_LIBDIR}/nest/lib{models,topology,random,...}.so``

Because ``pynestkernel.so`` does not go under the subdirectory ``nest``, ``${CMAKE_INSTALL_LIBDIR}`` refers to the highest level directory common to both.

This behaviour is consistent with the documentation (https://cmake.org/cmake/help/v3.0/module/GNUInstallDirs.html — see e.g. https://cmake.org/cmake/help/latest/command/install.html for example usage where they append a subdirectory to the CMake variable ``${CMAKE_INSTALL_INCLUDEDIR}``), so it should hold even if there were no ``pynestkernel.so`` and all dynamic libraries went into the ``nest`` subdirectory.

File: extras/nest-config.in (hunk 2 above)
Comment (clinssen) [reply extras/nest-config.in:69]: The behaviour is presently consistent: e.g. the MyModule cmake file takes the ``${CMAKE_INSTALL_LIBDIR}`` (equal to ``nest-config --libdir``) and again appends the ``nest`` subdirectory.

Comment (jougs) [reply extras/nest-config.in:69]: OK. Then maybe we should not change that. Thanks for the explanation.

**Pull request merged**